File: hierarchy-checker/src/main/java/com/yundin/hierarchy_checker/analyzer/ConcreteHierarchyAnalyzer.java
package com.yundin.hierarchy_checker.analyzer;

import android.view.View;
import android.view.ViewGroup;

import com.yundin.hierarchy_checker.adviser.Adviser;
import com.yundin.hierarchy_checker.outputter.RecommendationOutputter;

import org.jetbrains.annotations.NotNull;


public class ConcreteHierarchyAnalyzer implements HierarchyAnalyzer {

   private RecommendationOutputter outputter;
   private ViewGroup.OnHierarchyChangeListener hierarchyChangeListener;
   private Adviser adviser;

   public ConcreteHierarchyAnalyzer(RecommendationOutputter outputter, Adviser adviser) {
      this.outputter = outputter;
      this.hierarchyChangeListener = new HierarchyChangeListener();
      this.adviser = adviser;
   }

   @Override
   public void analyzeDynamicHierarchy(@NotNull View root) {
      proceedView(root);
      if (root instanceof ViewGroup) {
         ViewGroup viewGroup = (ViewGroup) root;
         viewGroup.setOnHierarchyChangeListener(hierarchyChangeListener);

         for (int i = 0; i < viewGroup.getChildCount(); i++) {
            analyzeDynamicHierarchy(viewGroup.getChildAt(i));
         }
      }
   }

   private void proceedView(@NotNull View view) {
      String alternative = adviser.findAlternative(view.getClass().getSimpleName());
      if (alternative != null) {
         outputter.output(view, alternative);
      }
   }

   class HierarchyChangeListener implements ViewGroup.OnHierarchyChangeListener {
      @Override
      public void onChildViewAdded(View parent, View child) {
         analyzeDynamicHierarchy(child);
      }

      @Override
      public void onChildViewRemoved(View parent, View child) {}
   }
}

File: hierarchy-checker/src/main/java/com/yundin/hierarchy_checker/analyzer/ConcreteHierarchyAnalyzer.kt
package com.yundin.hierarchy_checker.analyzer

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.OnHierarchyChangeListener
import androidx.core.view.ViewCompat
import com.yundin.hierarchy_checker.adviser.Adviser
import com.yundin.hierarchy_checker.outputter.RecommendationOutputter
import java.util.*

class ConcreteHierarchyAnalyzer(
    adviser: Adviser,
    outputter: RecommendationOutputter
) : HierarchyAnalyzer(adviser, outputter) {

    private val hierarchyChangeListener: OnHierarchyChangeListener = HierarchyChangeListener()
    private val known = hashSetOf<Int>()

    override fun analyzeDynamicHierarchy(root: View) {
        proceedView(root)
        if (root is ViewGroup) {
            root.setOnHierarchyChangeListener(hierarchyChangeListener)
            for (i in 0 until root.childCount) {
                analyzeDynamicHierarchy(root.getChildAt(i))
            }
        }
    }

    private fun proceedView(view: View) {
        if (view.id == -1) {
            view.id = ViewCompat.generateViewId()
        }
        if (!known.contains(view.id)) {
            known.add(view.id)
            adviser.findAlternativeAsync(view.javaClass.simpleName) { alternative ->
                outputter.output(view, alternative)
            }
        }
    }

    internal inner class HierarchyChangeListener : OnHierarchyChangeListener {

        override fun onChildViewAdded(parent: View, child: View) {
            analyzeDynamicHierarchy(child)
        }

        override fun onChildViewRemoved(parent: View, child: View) {}
    }
}